### PR TITLE
Fix test and AltTree

### PR DESCRIPTION
**Pull request by Mr-Leshiy**

### CODE REVIEW

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -40,11 +40,9 @@ add_library(${LIB_NAME} ${BUILD}
         $<TARGET_OBJECTS:sha256>
         $<TARGET_OBJECTS:bigdecimal>
         $<TARGET_OBJECTS:vblake>
+        $<TARGET_OBJECTS:secp256k1>
```
Comment (blackyblack) [inline src/CMakeLists.txt:43]: Why this change? It will block from linking with BTC.

File: src/blockchain/alt_block_tree.cpp [hunk 2]
```diff
@@ -173,76 +244,52 @@ void addContextToBlockIndex(BlockIndex<AltBlock>& index,
...
+
+  auto& vtbs = index.containingContext.top().vtbs;
+  auto vtbs_end = vtbs.end();
+  auto removeVTB = [&](const VTB& vtb) {
+    vtbs_end = std::remove(vtbs.begin(), vtbs_end, vtb);
   };
 
   // remove ATV context
-  std::for_each(p.atv.context.rbegin(), p.atv.context.rbegin(), remove);
+  std::for_each(p.atv.context.rbegin(), p.atv.context.rbegin(), removeBlock);
```
Comment (blackyblack) [inline src/blockchain/alt_block_tree.cpp:287]: Is it correct to have p.atv.context.rbegin() for both begin and end of the for_each?

File: src/blockchain/pop/vbk_block_tree.cpp [hunk 3]
```diff
@@ -106,11 +107,19 @@ bool VbkBlockTree::addPayloads(const VbkBlock& block,
 }
 
 void VbkBlockTree::removePayloads(const VbkBlock& block,
-                                  std::vector<payloads_t>& payloads) {
+                                  const std::vector<payloads_t>& payloads) {
   auto* index = getBlockIndex(block.getHash());
-  assert(index);  // must exist
+  removePayloads(index, payloads);
```
Comment (blackyblack) [inline src/blockchain/pop/vbk_block_tree.cpp:112]: I think it should be possible to automatically detect payloads containing in the block.

File: src/CMakeLists.txt (hunk 1 above)
Comment (Mr-Leshiy) [reply src/CMakeLists.txt:43]: I have changed it because locally with shared build, tests does not link to the seck256 symbols. So I have discussed with Bohdan and we decided to switch to the old variant

File: src/blockchain/alt_block_tree.cpp (hunk 2 above)
Comment (Mr-Leshiy) [reply src/blockchain/alt_block_tree.cpp:283]: fixed

File: src/blockchain/pop/vbk_block_tree.cpp (hunk 3 above)
Comment (Mr-Leshiy) [reply src/blockchain/pop/vbk_block_tree.cpp:112]: This method does not remove all payloads from the block, this method just remove payloads that are provided.

File: src/blockchain/alt_block_tree.cpp [hunk 4]
```diff
@@ -120,18 +124,55 @@ void AltTree::removePayloads(const AltBlock& containingBlock,
   assert(index);
 
   cmp_.removePayloads(*index, payloads);
+
+  if (index->containingContext.top().vtbs.empty() &&
+      index->containingContext.top().vtbs.empty()) {
```
[the remark this answers is not in the document]
Comment (Mr-Leshiy) [reply src/blockchain/alt_block_tree.cpp:125]: fixed

File: src/CMakeLists.txt (hunk 1 above)
Comment (Warchant) [reply src/CMakeLists.txt:43]: @blackyblack object libraries, as turned out, can not be linked to libs (source is docs)

File: include/veriblock/blockchain/alt_block_tree.hpp [hunk 5]
```diff
@@ -71,10 +71,6 @@ struct AltTree {
 
   VbkBlockTree& vbk() { return cmp_.getProtectingBlockTree(); }
   const VbkBlockTree& vbk() const { return cmp_.getProtectingBlockTree(); }
-  VbkBlockTree::BtcTree btc() { return cmp_.getProtectingBlockTree().btc(); }
-  const VbkBlockTree::BtcTree& btc() const {
-    return cmp_.getProtectingBlockTree().btc();
-  }
```
Comment (Warchant) [inline include/veriblock/blockchain/alt_block_tree.hpp:77]: why did you remove them?

File: src/blockchain/alt_block_tree.cpp [hunk 6]
```diff
@@ -120,18 +120,51 @@ void AltTree::removePayloads(const AltBlock& containingBlock,
   assert(index);
 
   cmp_.removePayloads(*index, payloads);
+
+  if (index->containingContext.back().vbk.empty() &&
+      index->containingContext.back().vtbs.empty()) {
```
Comment (Warchant) [inline src/blockchain/alt_block_tree.cpp:125]: add method to context, and use it here 
```
bool empty() const noexcept {
  return vbk.empty () && vtbs.empty();
}
```

File: src/blockchain/alt_block_tree.cpp [hunk 7]
```diff
@@ -120,18 +120,51 @@ void AltTree::removePayloads(const AltBlock& containingBlock,
...
-        auto& vbk = index.containingContext.vbk;
-        for (size_t i = 0, size = vbk.size(); i < size; i++) {
-          if (!tree().acceptBlock(vbk[i], state)) {
-            return state.addIndex(i).Invalid("bad-vbk-block");
+        if (index.containingContext.empty()) {
+          return true;
+        }
+
+        // step 1
+        for (const auto& b : index.containingContext.back().vbk) {
```
Comment (Warchant) [inline src/blockchain/alt_block_tree.cpp:140]: instead of using `index.containingContext.back()` every time, do a rename instead
```
auto& ctx = index.containingContext.back();
for(... ctx) {
```

File: src/blockchain/alt_block_tree.cpp [hunk 8]
```diff
@@ -146,16 +179,46 @@ void AltTree::PopForkComparator::sm_t::unapplyContext(
...
+    return;
+  }
+
+  auto check = [&](const VbkBlock& block) -> bool {
+    auto* index = tree().getBlockIndex(block.getHash());
+    return !index && index->containingContext.empty();
+  };
+
+  // step 1
+  for (const auto& b : index.containingContext.back().vbk) {
```
Comment (Warchant) [inline src/blockchain/alt_block_tree.cpp:192]: same here

File: src/CMakeLists.txt (hunk 1 above)
Comment (blackyblack) [reply src/CMakeLists.txt:43]: How did it work on my workstations? Both Linux and Windows linked and worked fine.

File: src/blockchain/alt_block_tree.cpp (hunk 7 above)
Comment (Mr-Leshiy) [reply src/blockchain/alt_block_tree.cpp:140]: fixed

File: src/blockchain/alt_block_tree.cpp (hunk 8 above)
Comment (Mr-Leshiy) [reply src/blockchain/alt_block_tree.cpp:192]: fixed

File: src/blockchain/alt_block_tree.cpp (hunk 6 above)
Comment (Mr-Leshiy) [reply src/blockchain/alt_block_tree.cpp]: fixed

File: include/veriblock/blockchain/alt_block_tree.hpp (hunk 5 above)
Comment (Mr-Leshiy) [reply include/veriblock/blockchain/alt_block_tree.hpp:77]: we have the same getter in the vbkTree

File: src/blockchain/alt_block_tree.cpp [hunk 9]
```diff
@@ -120,18 +120,50 @@ void AltTree::removePayloads(const AltBlock& containingBlock,
...
+          auto* containingIndex =
+              tree().getBlockIndex(vtb.containingBlock.getHash());
+          if (!containingIndex) {
+            if (!tree().acceptBlock(vtb.containingBlock, state)) {
+              return state.Invalid("alt-accept-block");
+            }
+          }
+          if (!containingIndex ||
+              containingIndex->containingEndorsements.find(
+                  BtcEndorsement::fromContainer(vtb).id) ==
```
Comment (Warchant) [inline src/blockchain/alt_block_tree.cpp:162]:
```suggestion
                  BtcEndorsement::getId(vtb)) ==
```

File: src/blockchain/alt_block_tree.cpp [hunk 10]
```diff
@@ -146,103 +178,111 @@ void AltTree::PopForkComparator::sm_t::unapplyContext(
...
+    }
+    addBlock(p.atv.containingBlock);
+  }
+
+  // step 2, process VTBs
+  for (const auto& vtb : p.vtbs) {
+    auto* temp = tree.getBlockIndex(vtb.getContainingBlock().getHash());
+
+    if (!temp || temp->containingEndorsements.find(
+                     BtcEndorsement::fromContainer(vtb).id) ==
```
Comment (Warchant) [inline src/blockchain/alt_block_tree.cpp:253]:
```suggestion
                     BtcEndorsement::getId(vtb)) ==
```
